### PR TITLE
cpu reduce mode did not need to broadcast params test=develop

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: python/paddle/fluid/compiler.py [hunk 1]
```diff
@@ -110,6 +120,8 @@ def with_data_parallel(self,
             self._exec_strategy = ExecutionStrategy()
         if self._build_strategy is None:
             self._build_strategy = BuildStrategy()
+        self._build_strategy.is_distribution = _is_pserver_mode(
```
Comment (panyx0718) [inline python/paddle/fluid/compiler.py]: what is `is_distribution` strategy?

Comment (panyx0718) [reply python/paddle/fluid/compiler.py]: @chengduoZH

Comment (jacquesqiao) [reply python/paddle/fluid/compiler.py]: means it use parameter server, will have send/recv in program

File: paddle/fluid/framework/details/build_strategy.cc [hunk 2]
```diff
@@ -133,12 +133,15 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
...
     if (strategy_.is_distribution_) {
+      VLOG(3) << "dist train mode";
       multi_devices_pass = AppendPass("dist_multi_devices_pass").get();
     } else {
       if (strategy.reduce_ == BuildStrategy::ReduceStrategy::kAllReduce) {
+        VLOG(3) << "allreduce mode";
         multi_devices_pass =
             AppendPass("allreduce_mode_multi_devices_pass").get();
       } else if (strategy.reduce_ == BuildStrategy::ReduceStrategy::kReduce) {
+        VLOG(3) << "reduce mode";
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/build_strategy.cc]: This log is too simple, others may not know what it means if just look at the output logs.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc [hunk 3]
```diff
@@ -925,9 +925,7 @@ void DistSSAGraphBuilder::InsertCollectiveOp(ir::Graph *result,
 }
 
 void DistSSAGraphBuilder::InsertPostprocessOps(ir::Graph *result) const {
-  if (need_broadcast_var_ ||
-      (UseGPU() &&
-       strategy_.reduce_ == BuildStrategy::ReduceStrategy::kReduce)) {
+  if (UseGPU() && strategy_.reduce_ == BuildStrategy::ReduceStrategy::kReduce) {
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/multi_devices_graph_pass.cc:928]: If `need_broadcast_var_` is not useful, it should be removed exactly.

File: python/paddle/fluid/compiler.py (hunk 1 above)
Comment (chengduoZH) [reply python/paddle/fluid/compiler.py]: It should be `is_distributed`.

Comment (chengduoZH) [reply python/paddle/fluid/compiler.py]: It should be `is_distributed`.

Comment (typhoonzero) [reply python/paddle/fluid/compiler.py]: maybe `or self._build_strategy.num_trainers > 1` should not be added here. `is_distribution` is only for pserver mode.

File: paddle/fluid/framework/details/multi_devices_graph_pass.cc (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/multi_devices_graph_pass.cc:928]: done

File: python/paddle/fluid/compiler.py (hunk 1 above)
Comment (jacquesqiao) [reply python/paddle/fluid/compiler.py]: removed

**Review by typhoonzero — APPROVED** [review]
LGTM

File: python/paddle/fluid/compiler.py [hunk 4]
```diff
@@ -34,6 +35,15 @@ def _place_obj(place):
     return p
 
 
+def _is_pserver_mode(main_program):
+    main = main_program if main_program \
+        else framework.default_main_program()
+    for op in main.global_block().ops:
+        if op.type in ["send", "recv"]:
```
Comment (panyx0718) [inline python/paddle/fluid/compiler.py:42]: this is very fragile. Can you use a better solution? tag program in transpiler?

File: python/paddle/fluid/compiler.py (hunk 1 above)
Comment (panyx0718) [reply python/paddle/fluid/compiler.py]: `is_distribution` is not a build strategy, maybe should get rid of it.

File: python/paddle/fluid/compiler.py [hunk 5]
```diff
@@ -110,6 +120,7 @@ def with_data_parallel(self,
             self._exec_strategy = ExecutionStrategy()
         if self._build_strategy is None:
             self._build_strategy = BuildStrategy()
+        self._build_strategy.is_distribution = _is_pserver_mode(self._program)
```
Comment (panyx0718) [inline python/paddle/fluid/compiler.py:123]: this should be reverted. pserver mode is not equal to is_distribution

**Comment by panyx0718** [comment]
Can we revert this PR and redo review?

**Pull request merged**